### PR TITLE
Fix decomp behaviour in export training IR

### PR DESCRIPTION
**Pull request by tugsbayasgalan**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #134657
* #135080
* __->__ #134801

Subset of changes in https://github.com/pytorch/pytorch/pull/132901, can't land the previous one because it is too complicated. Rest of the change will be implemented as follow up after export design meeting. This part just makes the training IR -> inference IR decomp to have the same path as normal export. 

Differential Revision: [D62000525](https://our.internmc.facebook.com/intern/diff/D62000525)

### CODE REVIEW

**Comment by tugsbayasgalan** [comment]
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D62000525).

**Comment by tugsbayasgalan** [comment]
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D62000525).

File: test/export/test_export.py [hunk 1]
```diff
@@ -1467,37 +1465,39 @@ def forward(self, p_conv_weight, p_conv_bias, p_conv1d_weight, p_conv1d_bias, b_
...
         )
 
         ep_has_convd = ep_has_convd.run_decompositions(
             decomp_table=None, _preserve_ops=[torch.ops.aten.conv2d.default]
         )
+
         self.assertExpectedInline(
             str(ep_has_convd.graph_module.code).strip(),
             """\
 def forward(self, p_conv_weight, p_conv_bias, p_conv1d_weight, p_conv1d_bias, b_linear_weight, b_linear_bias, x, y):
```
Comment (pianpwk) [inline test/export/test_export.py:1492]: lol ok this test case makes much more sense now

File: test/export/test_export.py [hunk 2]
```diff
@@ -1727,9 +1727,9 @@ def forward(self, x):
             """\
 def forward(self, p_linear_weight, p_linear_bias, b_buffer, x):
     add = torch.ops.aten.add.Tensor(b_buffer, 5);  b_buffer = None
-    t = torch.ops.aten.t.default(p_linear_weight);  p_linear_weight = None
-    addmm = torch.ops.aten.addmm.default(p_linear_bias, x, t);  p_linear_bias = x = t = None
-    sum_1 = torch.ops.aten.sum.default(add)
+    permute = torch.ops.aten.permute.default(p_linear_weight, [1, 0]);  p_linear_weight = None
+    addmm = torch.ops.aten.addmm.default(p_linear_bias, x, permute);  p_linear_bias = x = permute = None
+    sum_1 = torch.ops.aten.sum.dim_IntList(add, [])
```
Comment (pianpwk) [inline test/export/test_export.py:1732]: noob question: why the change here?

File: torch/_export/utils.py [hunk 3]
```diff
@@ -661,15 +876,16 @@ def _detect_fake_mode_from_gm(
...
-            "example_value" in node.meta or "val" in node.meta
-        ):
-            fake_val = None
-            if "example_value" in node.meta:
-                fake_val = node.meta["example_value"]
-            elif "val" in node.meta:
-                fake_val = node.meta["val"]
-            if fake_val is not None and isinstance(fake_val, torch.Tensor):
-                fake_vals.append(fake_val)
+
```
Comment (pianpwk) [inline torch/_export/utils.py:879]: hm could you explain this change?

Comment (pianpwk) [inline torch/_export/utils.py:879]: I think this makes sense, but could you explain what prompted the change? The previous implementation seemed equivalent, and only required one loop right?

File: torch/export/_trace.py [hunk 4]
```diff
@@ -1336,6 +1309,13 @@ def _strict_export_lower_to_aten_ir(
 
     _normalize_nn_module_stack(gm_torch_level, type(mod))
 
+    params_buffers_to_node_meta = _collect_param_buffer_metadata(gm_torch_level)
```
Comment (pianpwk) [inline torch/export/_trace.py:1312]: nice

File: torch/export/_trace.py [hunk 5]
```diff
@@ -1364,26 +1344,9 @@ def _strict_export_lower_to_aten_ir(
...
-                    for k, v in params_buffers_to_node_meta[param_name].items():
-                        node.meta[k] = v
-            if node.target in export_graph_signature.inputs_to_buffers:
-                buffer_name = export_graph_signature.inputs_to_buffers[node.target]
-                if buffer_name in params_buffers_to_node_meta:
-                    for k, v in params_buffers_to_node_meta[buffer_name].items():
-                        node.meta[k] = v
+    _populate_param_buffer_metadata_to_new_gm(
+        params_buffers_to_node_meta, gm, export_graph_signature
+    )
```
Comment (pianpwk) [inline torch/export/_trace.py:1349]: nice

File: torch/_export/utils.py [hunk 6]
```diff
@@ -42,6 +45,175 @@
 }
 
 
+def _collect_and_set_constant_attrs(
```
Comment (pianpwk) [inline torch/_export/utils.py:48]: nit: could you help add some context in the docstrings as to when/where these functions are used? e.g. in strict-mode, after dynamo export. Just makes it easier to parse through this file when a lot's going on.

File: torch/export/exported_program.py [hunk 7]
```diff
@@ -271,87 +280,14 @@ def _override_decomp_aten_to_variants():
         yield
 
 
-def _rename_without_collisions(
```
Comment (pianpwk) [inline torch/export/exported_program.py:274]: thanks, this was the wrong place for this in the 1st place

File: torch/export/exported_program.py [hunk 8]
```diff
@@ -655,6 +545,31 @@ def update_arg(old_arg, new_ph):
     return gm, new_graph_signature
 
 
+def _remove_unneccessary_copy_op_pass(
```
Comment (pianpwk) [inline torch/export/exported_program.py:548]: noob question: why is this op special?

**Comment by tugsbayasgalan** [comment]
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D62000525).

**Review by avikchaudhuri — APPROVED** [review]
lgtm after discussing offline

**Review by angelayi — APPROVED** [review]

**Comment by tugsbayasgalan** [comment]
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D62000525).
